### DISCUSSION TITLE
Use custom_jacobian option.

### DISCUSSION THREAD

**Discussion by charleskhazoom** [casadi/casadi#3245] [Q&A] [answered]
I was wondering how to use the `custom_jacobian` option in casadi 3.6.3. Here is my attempt with the following code, but it does not give the correct result.

```
x = casadi.MX.sym('x',2,1);
y = casadi.MX(2,1);
y(1) = sin(x(1));
y(2) = cos(x(2));

f = casadi.Function('f',{x},{y})
df1 = casadi.Function('df1',{x},{jacobian(f(x),x)})


dummy = casadi.MX.sym('dummy',f.size1_out(0))
jac_f_withdiff = casadi.Function('jac_f_withdiff',{x, dummy},{[x(1)^2, x(1)*x(2); x(2)^2, 2*x(1)*x(2)]})

options = struct("custom_jacobian", jac_f_withdiff, "jac_penalty",0, "always_inline",false, "never_inline",true)

% Create new function with custom jacobian
f_withdiff = casadi.Function('f_withdiff', {x}, {y}, options)

% Get custom jacobian from function
df1_custom = casadi.Function('df1_custom',{x},{jacobian(f_withdiff(x),x)})

% Evaluate the Jacobians
x0 = [2;3]
full(df1(x0))

full(df1_custom(x0))

full(jac_f_withdiff(x0,[]))
assert(full(df1_custom(x0)) == full(jac_f_withdiff(x0,[])))
```

**Comment by brunoml5** [comment]
Hi!

In the [release notes](https://web.casadi.org/get/), there is the following message:

> breaking custom_jacobian semantics changed. The Function must now return individual blocks (Jacobian of an output w.r.t. to an input)

I tried to interpret it and did some tests in matlab with your code, but I don't know what it means. Hopefully one of the developers answers it.

**Reply by charleskhazoom** [reply]
Right, I saw the release note too, and did not understand either.

**Comment by jaeandersson** [comment, marked as answer]
`Function::jacobian` used to return a single large Jacobian matrix corresponding to the Jacobian of all outputs with respect to all inputs. It now instead returns multiple Jacobian blocks. So for example for $f: (x, y) \rightarrow (f, g)$, you will now gets four separate blocks corresponding to df/dx, df/dy, dg/dx, dg/dy. The "custom_jacobian" has been updated accordingly, so if you provide your own Jacobian function, it should match the new syntax of `Function.jacobian()`.

If you are unsure about the syntax to use, just call `f.jacobian()` for an equivalent symbolic function and see what you get.
